[Solved #5, #6, #9] Setup state manager

supunTE opened the pull request:
Related #5

supunTE committed:
feat(data models): make seperated data models

Make seperated data models to manage transmitting data and in-app memory data.

BREAKING CHANGE: Seperate data models, generate nx lib

re #5

supunTE committed:
feat(redux): initialize redux store with rematchjs

Implement state management with redux rematchjs, Create store.

re #5

supunTE commented:
- [x] Initialize Redux Rematch
- [x] Make app data models
- [x] Initialize Store
- [x] Make reducers
- [x] Make effects
- [x] Create Utils
- [x] Error/Warn logger
- [x] Error logger into state

pasinducw commented on libs/models/src/lib/app/stateModel.ts:
```diff
@@ -0,0 +1,19 @@
+import { AppErrorLogs } from './errorLog';
```
use `kebab-case` for files unless they:
1. define and export a react component (Eg: `HomeScreen.tsx`)
2. define and export a single class (Eg: `HttpService.ts`)

supunTE replied:
Changed file names in 47ae59b714a45c2b1d3692f62afe5c19b4cdf1b3

supunTE committed:
feat(rematch): create rematch effects and reducers

Create reducer functions, effect functions. Create templates for tournament, race, player ids. Add
functionality to loading text, loading avatar.

re #5

supunTE committed:
feat(rematch): create rematch effects and reducers

Create endCountdown, endRace, sendTypeLog effect functions

re #5

supunTE committed:
feat(util): initialize util lib

re #5

supunTE committed:
refactor(util): move util functions from effects to seperated util files

Move genrateAvatarLink, generateUid, giveZeroPadding to utils. Create util file for calculateTimeout.

BREAKING CHANGE: Util functions are moved

re #5

supunTE committed:
feat: generate avatar link will now use playername as the seed

supunTE committed:
feat(rematch): create rematch effects and reducers

create reducer functions for updateTournament, updateLeaderboard, updatePlayer, removeTournament

re #5

supunTE committed:
feat(data models): add raceStartedBy property to the race model

Add raceStartedBy property to the race model in both app and data layers to record the player id who
started the race.

re #5

supunTE committed:
feat(effects): add updateRaceReducer to setReadyTournament effect

setReadyTournament effect will now update race data also.

re #5

supunTE committed:
feat(util): create extractIds util function

Create extractIds util function for extract piece of id from id collection.

re #5

supunTE committed:
fix(reducers): prevent updatePlayerLogReducer replacing new logs to previous

Make changes to prevent updatePlayerLogReducer from replacing previous logs.

re #5

supunTE committed:
feat(effects): make sendTypeLog to dispatch updatePlayerLogReducer

re #5

supunTE committed:
feat(util): add util function to calculate timeout timer

re #5

supunTE committed:
refactor(initialstate): populate initial state

supunTE committed:
feat(util): add functions to calculate wpm, elpased time, distance for leaderboard

re #5

supunTE committed:
feat(util): set functions to convert plyaer name to hex in generateAvatarLink

re #5

supunTE committed:
feat(effects): create generateLeaderboard effect

re #5

supunTE committed:
refactor(data models): change data model collection names

Change data model collection names to state and sockets

BREAKING CHANGE: Change folder names

re #5

supunTE committed:
refactor(store): organize store files

Organize store files; seperate reducers, effects and payloads into files

BREAKING CHANGE: Change file directory, new files added

re #5

supunTE committed:
test(reducers): add return types to reducers

supunTE committed:
feat(logger): upgrade logger function to emit warns and infos

supunTE committed:
feat(effects): add loggers to effects to raise errors

supunTE committed:
feat(logger): add max log limit and write function to delete oldest

re #5

supunTE committed:
style: change file names to kebab-case

supunTE commented:
Solved #5 #6 #9

supunTE committed:
fix(logger): fix logger raisers which raise logs for false cases

nadunys reviewed (approved):
Looks good to me

pasinducw commented on apps/client/src/main.tsx:
```diff
@@ -8,7 +10,9 @@ const root = ReactDOM.createRoot(
-    <App />
-  </StrictMode>,
+  <Provider store={store}>
+    <StrictMode>
```
Why is `Provider` outside of `StrictMode`?

supunTE replied:
Changed 95f0ac22e69d8ae243c63c4cd59f9fd38b67ab91

pasinducw commented on libs/models/src/lib/sockets/player.ts:
```diff
@@ -42,6 +42,7 @@ export interface Player {
 }
 
 export interface PlayerWithLogs extends Omit<Player, 'state' | 'lobbyId'> {
+  //rem lobbyId
```
What's this?

supunTE replied:
Removed unnecessary comments. 755e6524b5cc463844463d50060be3604003a6a2

pasinducw commented on libs/models/src/lib/state/leaderboard.ts:
```diff
@@ -0,0 +1,24 @@
+import { AppPlayerId } from './player';
```
add consistent spacing between the types

supunTE replied:
Add consistent spacing 755e6524b5cc463844463d50060be3604003a6a2

pasinducw commented on libs/models/src/lib/state/log-message.ts:
```diff
@@ -0,0 +1,28 @@
+export enum AppErrorCode {
```
Please add consistent spacing between the type definitions

supunTE replied:
Add consistent spacing 755e6524b5cc463844463d50060be3604003a6a2

pasinducw commented on libs/models/src/lib/state/player.ts:
```diff
@@ -0,0 +1,28 @@
+  Racing = 'racing',
+}
+
+export type AppTimestamp = number;
```
Add a jsdoc comment describing what this number represents

supunTE replied:
Updated docs da7c49d59ee51af3ed9dcea30e7055bd99311746

pasinducw commented on libs/models/src/lib/state/player.ts:
```diff
@@ -0,0 +1,28 @@
+import { AppRaceId } from './race';
```
Add consistent spacing between the type definitions
Eg: Empty line between each new type definition

supunTE replied:
Add consistent spacing 755e6524b5cc463844463d50060be3604003a6a2

pasinducw commented on libs/models/src/lib/state/race.ts:
```diff
@@ -0,0 +1,17 @@
+import { AppPlayerId, AppPlayer } from './player';
```
Please add consistent spacing between type definitions

supunTE replied:
Add consistent spacing 755e6524b5cc463844463d50060be3604003a6a2

pasinducw commented on libs/models/src/lib/state/state-model.ts:
```diff
@@ -0,0 +1,19 @@
+import { AppRaces } from './race';
+import { AppTournaments } from './tournament';
+
+export enum AppIdNumberType {
```
Why do we have `AppIdNumberType` defined within `state-model`?
Can we move it to its own file?

supunTE replied:
Moved aeb12cc93d2ac598fa64ebcccfc3c07d30251b2b

pasinducw commented on libs/store/src/lib/payloads/reducers.ts:
```diff
@@ -0,0 +1,55 @@
+} from '@razor/models';
+
+//reducer Playloads
+export type addTournamentReducerPayload = {
```
Please have all `types` in `PascalCase`; not `camelCase`.

supunTE replied:
changed in 7b3383b84c4b1723ebff6302c51e69b2535326f5 & 9d9abb3b34f8680fbfaadb6ffbcdaa1043ca0d28

pasinducw commented on libs/store/src/lib/effects/player.ts:
```diff
@@ -0,0 +1,112 @@
+  payload: joinPlayerPayload,
+  state: RootState,
+): Promise<void> => {
+  const { tid, playerName }: { tid: string; playerName: string } = payload;
```
Why do you need to explicitly state types for `tid` and `payerName` here?

supunTE replied:
Fixed in PR #14 under ab67cf8d4a757a74236342dbdc0eb9475ec389fe

pasinducw commented on libs/store/src/lib/effects/player.ts:
```diff
@@ -0,0 +1,112 @@
+  const { tid, playerName }: { tid: string; playerName: string } = payload;
+  let tournamentId: string;
+
+  if (playerName.length < 2 || playerName.length > 10)
```
Why are we not terminating the the execution of the effect when the constraints fail?

supunTE replied:
Fix in PR #14 under commit 2c90ad8bcbf7254f6da140cca1193fcea40da26e

pasinducw commented on libs/store/src/lib/effects/player.ts:
```diff
@@ -0,0 +1,112 @@
+): Promise<void> => {
+  const { playerId }: { playerId: AppPlayerId } = payload;
+
+  if (!playerId) playerNotFound(dispatch, playerId);
```
Why do we continue execution despite having failed a check?

supunTE replied:
Fix in PR #14 under commit 2c90ad8bcbf7254f6da140cca1193fcea40da26e

pasinducw commented on libs/store/src/lib/effects/player.ts:
```diff
@@ -0,0 +1,112 @@
+
+  const playerLogId: AppPlayerLogId = `${raceId}-${playerId}`;
+
+  if (!playerId) playerNotFound(dispatch, playerId);
```
Why do we continue execution despite having failed an assertion?

supunTE replied:
Fix in PR #14 under commit 2c90ad8bcbf7254f6da140cca1193fcea40da26e

pasinducw commented on libs/store/src/lib/effects/player.ts:
```diff
@@ -0,0 +1,112 @@
+  const playerLogId: AppPlayerLogId = `${raceId}-${playerId}`;
+
+  if (!playerId) playerNotFound(dispatch, playerId);
+  if (!raceId) raceNotFound(dispatch, raceId);
```
Why do we continue execution despite having failed an assertion?

supunTE replied:
Fix in PR #14 under commit 2c90ad8bcbf7254f6da140cca1193fcea40da26e

pasinducw commented on libs/store/src/lib/effects/race.ts:
```diff
@@ -0,0 +1,168 @@
+  }: { tournamentId: AppTournamentId; playerId: AppPlayerId } = payload;
+
+  if (!state.game.tournamentsModel[tournamentId])
+    tournamentNotFound(dispatch, tournamentId, `Started by: ${playerId}`);
```
Why do we continue execution despite having failed an assertion?

supunTE replied:
Fix in PR #14 under commit a89f04352fcf13545275e01bbd41b93a1aaa26ac

pasinducw commented on libs/store/src/lib/effects/race.ts:
```diff
@@ -0,0 +1,168 @@
+  if (!state.game.tournamentsModel[tournamentId])
+    tournamentNotFound(dispatch, tournamentId, `Started by: ${playerId}`);
+  if (!state.game.playersModel[playerId])
+    playerNotFound(
```
Why do we continue execution despite having failed an assertion?

supunTE replied:
Fix in PR #14 under commit a89f04352fcf13545275e01bbd41b93a1aaa26ac

pasinducw commented on libs/store/src/lib/effects/race.ts:
```diff
@@ -0,0 +1,168 @@
+    const player = state.game.playersModel[id];
+    if (!player) {
+      if (!state.game.playersModel[playerId])
+        playerNotFound(
```
Why do we continue execution despite having failed an assertion?

supunTE replied:
Fix in PR #14 under commit a89f04352fcf13545275e01bbd41b93a1aaa26ac

pasinducw commented on libs/store/src/lib/effects/race.ts:
```diff
@@ -0,0 +1,168 @@
+    state.game.tournamentsModel[tournamentId].raceIds.length;
+  const raceIndex = numberOfRacesBefore
+    ? giveZeroPadding(numberOfRacesBefore.toString(), 3)
+    : '000';
```
what's this one doing?

supunTE replied:
Fix in PR #14 under commit 027ab37b45b045ae9328637779ce152f7989e5dd

pasinducw commented on libs/store/src/lib/effects/race.ts:
```diff
@@ -0,0 +1,168 @@
+  for (const id of state.game.tournamentsModel[tournamentId].playerIds) {
+    const player = state.game.playersModel[id];
+    if (!player) {
+      if (!state.game.playersModel[playerId])
```
What happens in the `else` scenario of this?

supunTE replied:
Fixed in PR #14 under commit a89f04352fcf13545275e01bbd41b93a1aaa26ac on line 77

pasinducw commented on libs/store/src/lib/effects/race.ts:
```diff
@@ -0,0 +1,168 @@
+    }
+  }
+
+  const raceText = await loadRacingText();
```
I think that the racingText must come as an input payload for this effect; because, including the text loading logic here couples our game with the text provider.

The text provider would ideally live in the backend; having no connection to the game state, or its definitions.
WDYT?

supunTE replied:
Yes, this was changed in PR #14 under commit 4cc6db0746f129bb874452203467d069dfa797e0 and a89f04352fcf13545275e01bbd41b93a1aaa26ac

pasinducw commented on libs/store/src/lib/effects/race.ts:
```diff
@@ -0,0 +1,168 @@
+): Promise<void> => {
+  const { tournamentId }: { tournamentId: AppTournamentId } = payload;
+
+  if (!state.game.tournamentsModel[tournamentId])
```
Why do we continue execution despite having failed an assertion?

supunTE replied:
Fix in PR #14 under commit a89f04352fcf13545275e01bbd41b93a1aaa26ac

pasinducw commented on libs/store/src/lib/effects/race.ts:
```diff
@@ -0,0 +1,168 @@
+  const { raceId }: { raceId: AppRaceId } = payload;
+
+  if (!state.game.racesModel[raceId])
+    raceNotFound(dispatch, raceId, `While race ending`);
```
Why do we continue execution despite having failed an assertion?

supunTE replied:
Fix in PR #14 under commit a89f04352fcf13545275e01bbd41b93a1aaa26ac

pasinducw commented on libs/store/src/lib/effects/tournament.ts:
```diff
@@ -0,0 +1,29 @@
+  const { tournamentId }: { tournamentId: AppTournamentId } = payload;
+
+  if (!state.game.tournamentsModel[tournamentId])
+    tournamentNotFound(dispatch, tournamentId, `While setting ready`);
```
Why do we continue execution despite having failed an assertion?

supunTE replied:
Fix in PR #14 under commit 6a024d6d4259d2be4fff29e45608b3a6b3735f75

pasinducw commented on libs/store/src/lib/loggers/player.ts:
```diff
@@ -0,0 +1,36 @@
+import { AppErrorCode, AppMessageLogType } from '@razor/models';
+import { Dispatch } from '../store';
+
+export const playerNotFound = async (
```
why do we have async functions here?
If `dispatch.games.sendLogMessage` is asynchronous, we need to `await` on the response.

supunTE replied:
Fixed in PR #14 under commit 563c7da7291336f0b666f5c471e3d5046054c636

pasinducw commented on libs/store/src/lib/loggers/player.ts:
```diff
@@ -0,0 +1,36 @@
+    related: additionalMessage ? additionalMessage : '',
+    type: AppMessageLogType.Error,
+  });
+  return;
```
`return` at the end of the function is redundant. The function is going to anyway return `undefined.

supunTE replied:
Fixed in PR #14 under commit a30c31abb7742dfe160b204a5bca3285d6a0d65a

pasinducw commented on libs/store/src/lib/loggers/player.ts:
```diff
@@ -0,0 +1,36 @@
+    related: '',
+    type: AppMessageLogType.Error,
+  });
+  return;
```
`return` at the end of the function is redundant. The function is going to anyway return `undefined`.

supunTE replied:
Fixed in PR #14 under commit a30c31abb7742dfe160b204a5bca3285d6a0d65a

pasinducw commented on libs/store/src/lib/loggers/player.ts:
```diff
@@ -0,0 +1,36 @@
+    related: '',
+    type: AppMessageLogType.Error,
+  });
+  return;
```
`return` at the end of the function is redundant. The function is going to anyway return `undefined`.

supunTE replied:
Fixed in PR #14 under commit a30c31abb7742dfe160b204a5bca3285d6a0d65a

pasinducw commented on libs/store/src/lib/loggers/player.ts:
```diff
@@ -0,0 +1,36 @@
+};
+
+export const invalidPlayerNameLength = async (dispatch: Dispatch) => {
+  dispatch.game.sendLogMessage({
```
Do we need to `await` on the result?
If not, why do we have specified an `async` function?

supunTE replied:
Fixed in PR #14 under commit 563c7da7291336f0b666f5c471e3d5046054c636

pasinducw commented on libs/store/src/lib/loggers/player.ts:
```diff
@@ -0,0 +1,36 @@
+};
+
+export const invalidPlayerName = async (dispatch: Dispatch) => {
+  dispatch.game.sendLogMessage({
```
Do we need to `await` on the result?
If not, why do we have specified an `async` function?

supunTE replied:
Fixed in PR #14 under commit 563c7da7291336f0b666f5c471e3d5046054c636

pasinducw commented on libs/store/src/lib/loggers/race.ts:
```diff
@@ -0,0 +1,16 @@
+  rid: string,
+  additionalMessage?: string,
+) => {
+  dispatch.game.sendLogMessage({
```
Do we need to `await` on the result?
If not, why do we have specified an `async` function?

supunTE replied:
Fixed in PR #14 under commit 563c7da7291336f0b666f5c471e3d5046054c636

pasinducw commented on libs/store/src/lib/loggers/race.ts:
```diff
@@ -0,0 +1,16 @@
+    related: additionalMessage ? additionalMessage : '',
+    type: AppMessageLogType.Error,
+  });
+  return;
```
this statement is redundant.

supunTE replied:
Fixed in PR #14 under commit a30c31abb7742dfe160b204a5bca3285d6a0d65a

pasinducw commented on libs/store/src/lib/loggers/tournament.ts:
```diff
@@ -0,0 +1,16 @@
+  tid: string,
+  additionalMessage?: string,
+) => {
+  dispatch.game.sendLogMessage({
```
Do we need to `await` on the result?
If not, why do we have specified an `async` function?

supunTE replied:
Fixed in PR #14 under commit 563c7da7291336f0b666f5c471e3d5046054c636

pasinducw commented on libs/store/src/lib/loggers/tournament.ts:
```diff
@@ -0,0 +1,16 @@
+    related: additionalMessage ? additionalMessage : '',
+    type: AppMessageLogType.Error,
+  });
+  return;
```
This statement is redundant.

supunTE replied:
Fixed in PR #14 under commit a30c31abb7742dfe160b204a5bca3285d6a0d65a

pasinducw commented on libs/store/src/lib/payloads/effects.ts:
```diff
@@ -0,0 +1,34 @@
+  AppTournamentId,
+} from '@razor/models';
+
+//effectsPayloads
```
Please use `PascalCase` in types.

supunTE replied:
changed in 7b3383b84c4b1723ebff6302c51e69b2535326f5 & 9d9abb3b34f8680fbfaadb6ffbcdaa1043ca0d28

pasinducw commented on libs/store/src/lib/reducers/add.ts:
```diff
@@ -0,0 +1,83 @@
+    raceId,
+    extractIdType.race,
+    extractIdType.tournament,
+  ) as AppTournamentId;
```
Can you write the function to return the relevant type based on the parameters passed in, instead of doing the type cast here?

supunTE replied:
Fixed in PR #14 under commit 42e5c196dd4d85bfe7c232dd718c2b5c5935231e

pasinducw commented on libs/store/src/lib/reducers/add.ts:
```diff
@@ -0,0 +1,83 @@
+    extractIdType.tournament,
+  ) as AppTournamentId;
+
+  if (raceId) {
```
Do we expect `raceId` to be undefined?

supunTE replied:
Fixed in PR #14 under commit 39ef350e9fa663424ef9b78c7f25a8c240263c36

pasinducw commented on libs/store/src/lib/reducers/add.ts:
```diff
@@ -0,0 +1,83 @@
+      [tournamentId]: {
+        ...state.tournamentsModel[tournamentId],
+        playerIds: [
+          ...(state.tournamentsModel[tournamentId] &&
```
You can simplify this statement as below:
`state.tournamentsModel[tournamentId]?.playerIds || []`

supunTE replied:
This logic is no longer used because functions are checking tournament (or any other related value) availability at the top of the function. 39ef350e9fa663424ef9b78c7f25a8c240263c36

pasinducw commented on libs/store/src/lib/reducers/logger.ts:
```diff
@@ -0,0 +1,33 @@
+    logModel = { ...newLogModel };
+  }
+
+  console.log(state.errorLogsModel);
```
You can't do console logs within reducers. 
A reducer needs to be a pure function, without side effects.
console.log is a side effect. Hence must be done within an effect.

supunTE replied:
This is removed in PR #14 https://github.com/the-ai-team/razor/pull/14/files#diff-65c47b8507de32a7a9165c89e550a79ba994cff94ce4489d5cbc7ba257ddcfb9

pasinducw commented on libs/store/src/lib/reducers/remove.ts:
```diff
@@ -0,0 +1,96 @@
+      ...state.tournamentsModel,
+      [tournamentId]: {
+        ...state.tournamentsModel[tournamentId],
+        playerIds: [
```
you can simplify this as below:
`state.tournamentsModel[tournamentId].playerIds.filter((id) => id !== playerId)`

supunTE replied:
simplified b6085c6c6aa6ad83c276be895e541d9b685b0414

pasinducw commented on libs/store/src/lib/reducers/remove.ts:
```diff
@@ -0,0 +1,96 @@
+  return newState;
+};
+
+export const removeTournamentReducer = (
```
Since your algorithm here includes multiple steps, please add a small comment before each code stub; so a person reading would easily understand what's being done

pasinducw replied:
Eg: `// get all player ids of all races in tournament`

supunTE replied:
Completed docs in 38e16c224293b3eabf0be79a998e2ba44a2b93fb

pasinducw commented on libs/store/src/lib/reducers/update.ts:
```diff
@@ -0,0 +1,79 @@
+    playerLogsModel: {
+      ...state.playerLogsModel,
+      [playerLogId]: [
+        ...(state.playerLogsModel[playerLogId]
```
You can simplify this as below:
`state.playerLogsModel[playerLogId] || []`

supunTE replied:
Fixed in PR #14 under commit 39ef350e9fa663424ef9b78c7f25a8c240263c36

pasinducw commented on libs/util/src/lib/calculate-timeout-timer.ts:
```diff
@@ -0,0 +1,7 @@
+const averageWPM = 50;
+
+export const calculateTimeoutTimer = async (text: string) => {
+  const averageWordLength = text.length / 5;
```
Can you please document the reason behind 5?
We don't want to have magic numbers

supunTE replied:
Completed docs  in PR #14 under commit 4d6c57dec3d53ee4b0ea4ee596fa7368c3a0833d

pasinducw commented on libs/util/src/lib/calculate-timeout-timer.ts:
```diff
@@ -0,0 +1,7 @@
+const averageWPM = 50;
+
+export const calculateTimeoutTimer = async (text: string) => {
```
Can you please add a jsdoc explanation for this? 
What's the unit of `timeoutDuration`?

supunTE replied:
Completed docs  in PR #14 under commit 4d6c57dec3d53ee4b0ea4ee596fa7368c3a0833d

pasinducw commented on libs/util/src/lib/extractIds.ts:
```diff
@@ -0,0 +1,44 @@
+export enum extractIdType {
```
Please use `PascalCase` for all types

supunTE replied:
Changed in #14 under commit ef2a597785f6a6f2133243b6e7423cd2d58adbde

pasinducw commented on libs/util/src/lib/extractIds.ts:
```diff
@@ -0,0 +1,44 @@
+export enum extractIdType {
+  tournament = 'tournament',
```
Please use `PascalCase` for enum identifiers.

supunTE replied:
Changed in #14 under commit ef2a597785f6a6f2133243b6e7423cd2d58adbde

pasinducw commented on libs/util/src/lib/extractIds.ts:
```diff
@@ -0,0 +1,44 @@
+  playerLog = 'playerLog',
+}
+
+export const extractId = (
```
Please do the type casts to the returned type, so that the consumers of this function will not have to do manual type casts.

supunTE replied:
Solved in PR #14 under commit 42e5c196dd4d85bfe7c232dd718c2b5c5935231e

pasinducw commented on libs/util/src/lib/generate-leaderboard.ts:
```diff
@@ -0,0 +1,141 @@
+} from '@razor/models';
+import { extractId, extractIdType } from './extractIds';
+
+export const generateLeaderboard = (
```
Can you please add short comments before each code stub, to describe what they do?

supunTE replied:
Completed docs  in PR #14 under commit 4d6c57dec3d53ee4b0ea4ee596fa7368c3a0833d

pasinducw commented on libs/util/src/lib/generate-leaderboard.ts:
```diff
@@ -0,0 +1,141 @@
+  return leaderboardEntries;
+};
+
+const calculateWPM = (length: number, logs: AppPlayerLog[]): number => {
```
Can you please add documentation about this algorithm?

supunTE replied:
Completed docs  in PR #14 under commit 4d6c57dec3d53ee4b0ea4ee596fa7368c3a0833d

pasinducw commented on libs/util/src/lib/give-zero-padding.ts:
```diff
@@ -0,0 +1,4 @@
+export const giveZeroPadding = (num: string, size: number): string => {
+  const s = num + '';
+  return s.padStart(size, '0');
```
You can simplify this as follows:
``return `${num}`.padStart(size, '0');``

supunTE replied:
Done a7a004fa2658f128a28d7d4735d943f10b8a030d

pasinducw commented on libs/util/src/lib/generateUid.ts:
```diff
@@ -0,0 +1,22 @@
+const playerIdLength = 8;
+const generalIdLength = 8;
+
+export const generateUid = async (type: AppIdNumberType): Promise<string> => {
```
have a dynamic return type based on the `type`

supunTE replied:
Solved in PR #14 under commit 0fbe5ff4d06d17be7e2cfbbc1cd667bd024e6cca

pasinducw reviewed (changes requested):
Added comments

supunTE committed:
fix(client): move store provider to inside of the strict mode

re #5

supunTE committed:
style(data models): add consistent spacing

re #5

supunTE committed:
docs(data models): update js docs

re #5

supunTE committed:
feat(data models): seperate id number type from state-model

Move AppIdNumberType from state-model to uid in state data models

re #5

supunTE committed:
refactor(payloads): change case of the type names to PascalCase in reducer payloads

re #5

supunTE committed:
refactor(payloads): change case of the type names to PascalCase in effect payloads

re #5

supunTE committed:
feat(util): simplify util give zero padding

re #5

supunTE committed:
refactor(reducers/remove): simplify logic using filter method
